Minor cleanups and some new example programs.

phbcanada opened the pull request:
The threaded streaming example program wasn't working so I fixed it.
The new "set_ftp" example is useful for me since I use FTP to record all my camera motion videos.
The "snap" example is useful for grabbing snapshots from each camera.
I use hostnames instead of IP addresses for my cameras. Just put the IP addresses in /etc/hosts.

phbcanada committed:
Update setup.py

Loosen restrictions.

phbcanada committed:
Update streaming_video.py

Improvements to share common code.
Fix threaded version to open window in main thread.

phbcanada committed:
Update stream.py

Forgot to commit with previous.
New method: stop_stream.

phbcanada committed:
Add check for login failure.

New camera method: is_logged_in

phbcanada committed:
New api method "set_network_ftp"

New example program to set ftp parameters.

phbcanada committed:
Two new example programs

get_info calls all the info methods.
snap will snap a single frame image.

phbcanada committed:
Update api_handler.py

Getting occasional garbage back from the cameras which causes the json conversion to crap out. Print some info when this happens.

Benehiko reviewed (changes requested):
Hey @phbcanada 

Thank you for the contribution! This looks quite good and we should easily be able to merge it :)

Just one question about the version specifier inside the `setup.py` file.

Benehiko commented:
Could we also use conventional commits on the PR title?

I know this repository doesn't have any real guidelines on this and in the past it's been quite lax, but I think it is good practice.

you can change the title to something like `fix: minor cleanups and new examples`.

https://www.conventionalcommits.org/en/v1.0.0/#summary

phbcanada commented on setup.py:
```diff
@@ -32,9 +32,9 @@ def find_version(*file_paths):
-        'numpy==1.19.4',
-        'opencv-python==4.4.0.46',
-        'Pillow==8.0.1',
+        'numpy>=1.19.4',
+        'opencv-python>=4.4.0.46',
+        'Pillow>=8.0.1',
     ],
```
The install failed on one of my systems so this was the easy fix. Not a big issue for me.

phbcanada commented:
I haven't used git/github for several years. Not really up on current conventions. Sorry.

Benehiko commented:
> I haven't used git/github for several years. Not really up on current conventions. Sorry.

np :)

phbcanada committed:
Add support for newer API version on some cameras.

The GetAbility command returns a property named "scheduleVersion" which indicates which API calls should be used. The new ones have a "V20" tacked onto the command name.

phbcanada committed:
Merge branch 'master' of github.com:phbcanada/reolinkapipy

tinito commented:
Any chances to get this merged? The set_network_ftp() method would be really helpful.

themoosman commented:
> Any chances to get this merged? The set_network_ftp() method would be really helpful.

There looks to be a simple requested change that still needs to be done.

phbcanada committed:
Add some support for V2 API. Misc fixes and additions.

Improve set_osd call.
Extend FTP setting support.
Add set_network_ntp
Add set_network_email
Fixes to add_user and modify_user.

phbcanada committed:
Correct call is actually GetMdAlarm

GetAlarm is not supported on newer cameras.

phbcanada committed:
Fixes to set_recording_encoding

Correct call is "SetEnc" not "SetRec".
Argument types changed.

Benehiko commented on reolinkapi/mixins/display.py:
```diff
@@ -38,18 +39,21 @@ def set_osd(self, bg_color: bool = 0, channel: float = 0, osd_channel_enabled: b
         """
-        body = [{"cmd": "SetOsd", "action": 1,
+        body = [{"cmd": "SetOsd", # "action": 1,
                  "param": {
                     "Osd": {
-                        "bgcolor": bg_color,
+                        # "bgcolor": bg_color,
```
why are these commented out?

Benehiko commented on reolinkapi/mixins/display.py:
```diff
@@ -38,18 +39,21 @@ def set_osd(self, bg_color: bool = 0, channel: float = 0, osd_channel_enabled: b
+                                    "pos": osd_time_pos},
+                        "watermark": int(osd_watermark_enabled),
                     }}}]
+        # print("SetOsd:", json.dumps(body[0], indent=4))
```
maybe just delete?

Benehiko commented on reolinkapi/mixins/network.py:
```diff
@@ -25,9 +25,127 @@ def set_net_port(self, http_port: float = 80, https_port: float = 443, media_por
         }}}]
         self._execute_command('SetNetPort', body, multi=True)
-        print("Successfully Set Network Ports")
+        # print("Successfully Set Network Ports")
```
```suggestion
        print("Successfully Set Network Ports")
```

Benehiko commented on reolinkapi/mixins/user.py:
```diff
@@ -32,23 +33,37 @@ def add_user(self, username: str, password: str, level: str = "guest") -> bool:
+                ]
         r_data = self._execute_command('ModifyUser', body)[0]
-        if r_data["value"]["rspCode"] == 200:
+        # print(f"modify user: {username}\nCamera responded with: {json.dumps(r_data, indent=4)}")
```
```suggestion
```

Benehiko commented on reolinkapi/handlers/api_handler.py:
```diff
@@ -64,28 +68,47 @@ def login(self) -> bool:
             param = {"cmd": "Login", "token": "null"}
             response = Request.post(self.url, data=body, params=param)
             if response is not None:
+                # print("LOGIN GOT: ", response.text)
```
```suggestion
```

Benehiko commented on reolinkapi/handlers/api_handler.py:
```diff
@@ -64,28 +68,47 @@ def login(self) -> bool:
         """
         try:
+            # Suppress only the single warning from urllib3 needed.
+            requests.packages.urllib3.disable_warnings(category=InsecureRequestWarning)
```
maybe make this optional?

or what about just leaving this to the client to set this as an env variable

```shell
PYTHONWARNINGS="ignore:Unverified HTTPS request"
```

Benehiko commented on reolinkapi/handlers/api_handler.py:
```diff
@@ -64,28 +68,47 @@ def login(self) -> bool:
-                    return True
-                print(self.token)
-                return False
+                    # print("Login success")
+                else:
+                    # print(self.token)
```
remove the comments or restore the print of "Login success?"

Benehiko reviewed (changes requested):
Looks good, just some minor changes. The `setup.py` is a bit concerning, IMO it needs to be fixed versions.

phbcanada committed:
Changes that were sitting here for a long time.

phbcanada committed:
Merge new get/set_device_name. Fix duplicate get_time error.
